integrationtest/performance/engine-elasticsearch/src/main/java/org/hibernate/search/engineperformance/elasticsearch/ChangesetGenerator.java

```diff
@@ -0,0 +1,191 @@
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.hibernate.search.engineperformance.elasticsearch.setuputilities.SearchIntegratorCreation;
+import org.hibernate.search.exception.AssertionFailure;
+import org.hibernate.search.util.impl.CollectionHelper;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.infra.ThreadParams;
+
+/**
+ * Holds the list of documents currently present in the index
+ * <p>
+ * Note that we cannot reuse IDs during an iteration, for instance
+ * adding a document for an ID that just got deleted,
+ * because changesets may be executed asynchronously and out of order.
+ * That's why we store consumed IDs, and we flip the add and delete sets
+ * from one iteration to another.
+ */
+@State(Scope.Thread)
+public class ChangesetGenerator {
+
+	private static final long SEED = 3210140441369L;
+
+	private int changesets;
+	private int addDeletesPerInvocation;
+	private int updatesPerInvocation;
+	private int threadIdIntervalFirst;
+
+	private BitSet toAdd;
+	private BitSet consumedToAdd;
+
+	private BitSet toDelete;
+	private BitSet consumedToDelete;
+
+	private BitSet toUpdate;
+	private BitSet consumedToUpdate;
+
+	@Setup(Level.Trial)
+	public void setup(EngineHolder eh, ThreadParams threadParams) {
+		int initialIndexSize = eh.getInitialIndexSize();
+		int addDeletesPerChangeset = eh.getAddsDeletesPerChangeset();
+		int updatesPerChangeset = eh.getUpdatesPerChangeset();
+		changesets = eh.getChangesetsPerFlush();
+
+		int threadIndex = threadParams.getSubgroupThreadIndex();
+
+		addDeletesPerInvocation = changesets * addDeletesPerChangeset;
+		updatesPerInvocation = changesets * updatesPerChangeset;
+		int threadIdIntervalSize = addDeletesPerInvocation + addDeletesPerInvocation + updatesPerInvocation;
+		threadIdIntervalFirst = initialIndexSize + threadIndex * threadIdIntervalSize;
+
+		toAdd = new BitSet( threadIdIntervalSize );
+		consumedToAdd = new BitSet( threadIdIntervalSize );
+		toDelete = new BitSet( threadIdIntervalSize );
+		consumedToDelete = new BitSet( threadIdIntervalSize );
+		toUpdate = new BitSet( threadIdIntervalSize );
+		consumedToUpdate = new BitSet( threadIdIntervalSize );
+
+		// Initialize the ID sets
+		List<Integer> shuffledIds = createShuffledIndexList( threadIdIntervalSize );
+		int deletesShuffledIdsOffset = addDeletesPerInvocation;
+		int updatesShuffledIdsOffset = deletesShuffledIdsOffset + addDeletesPerInvocation;
+		for ( int i = 0 ; i < addDeletesPerInvocation ; ++i ) {
+			toAdd.set( shuffledIds.get( i ) );
+		}
+		for ( int i = deletesShuffledIdsOffset ; i < updatesShuffledIdsOffset ; ++i ) {
+			toDelete.set( shuffledIds.get( i ) );
+		}
+		for ( int i = updatesShuffledIdsOffset ; i < threadIdIntervalSize ; ++i ) {
+			toUpdate.set( shuffledIds.get( i ) );
+		}
```

Comment:
I'm not sure I understand the goal here. How do you guarantee that `toAdd` and `toDelete` are disjoint sets ?